documentation/Config.md
# Overview
This page describes how to add in your own `attributes` to your service.

## Generation
Apon creation of your service, a `config` module is created with commented out templated code that looks as follows:


    package com.acme.config
    
    import com.twitter.app.{GlobalFlag, GlobalFlagVisible}
    import com.deciphernow.server.Implicits.stringOptionFlaggable
    
    //object anObjectFlag extends GlobalFlag[String]("HasSomeValue","Put description here.")
    
    //package somePackage {
    //  object anotherObjectFlag  extends GlobalFlag[String]("AnotherValue", "Put description here.")
    //}
    
    
### Example of adding capability
We are going to add an attribute to our newly created __myfirstmicroserivce__.   Edit __com.acme.config.Config__ and add __object version__, see below:


    package com.acme.config
    
    import com.twitter.app.{GlobalFlag, GlobalFlagVisible}
    import com.deciphernow.server.Implicits.stringOptionFlaggable
    import com.deciphernow.server.config.rest
    
    object version extends GlobalFlag[String]("0.0.0","Unassigned version returns 0.0.0")


### Scala reference
In the business module, lets edit __src/main/scala/com.acme.MyFirstMicroserviceManager__ and modify the __getPong__ method to the following:

    package com.acme
    
    /**
     *
     */
    class MyFirstMicroserviceManager {
    
      /**
        * Keep this. Use to see if the service is accepting connections.
        * @return
        */
      def getPong = {
        println("My added attribute : " + com.acme.config.version.apply)
        "pong\n"
      }
    }
 
Compile your service, fix the user issue, start it, hit the ping end-point. The print statement is going to echo the default value of __version__ to the output when you started the service with __console__. This is because we did __NOT__ assign a value in __parameters.config__.
[Reference Creating New Microservice if you forgot](CreatingNewMS.md)

    
    jvm 1    | 15:02:24.994 [WrapperSimpleAppMain] INFO  c.t.finatra.http.routing.HttpRouter - Adding routes
    jvm 1    | GET     /ping
    jvm 1    | 15:02:24.997 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer -  => com.twitter.server.handler.AdminRedirectHandler
    jvm 1    | 15:02:24.997 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin => com.twitter.server.handler.SummaryHandler
    jvm 1    | 15:02:24.997 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/server_info => com.twitter.finagle.Filter$$anon$1
    jvm 1    | 15:02:24.997 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/contention => com.twitter.finagle.Filter$$anon$1
    jvm 1    | 15:02:24.998 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/lint => com.twitter.server.handler.LintHandler
    jvm 1    | 15:02:24.998 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/lint.json => com.twitter.server.handler.LintHandler
    jvm 1    | 15:02:24.998 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/failedlint => com.twitter.server.handler.FailedLintRuleHandler
    jvm 1    | 15:02:24.998 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/threads => com.twitter.server.handler.ThreadsHandler
    jvm 1    | 15:02:24.998 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/threads.json => com.twitter.server.handler.ThreadsHandler
    jvm 1    | 15:02:24.998 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/announcer => com.twitter.finagle.Filter$$anon$1
    jvm 1    | 15:02:24.998 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/dtab => com.twitter.finagle.Filter$$anon$1
    jvm 1    | 15:02:24.998 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/pprof/heap => com.twitter.server.handler.HeapResourceHandler
    jvm 1    | 15:02:24.999 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/pprof/profile => com.twitter.server.handler.ProfileResourceHandler
    jvm 1    | 15:02:24.999 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/pprof/contention => com.twitter.server.handler.ProfileResourceHandler
    jvm 1    | 15:02:24.999 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/ping => com.twitter.server.handler.ReplyHandler
    jvm 1    | 15:02:24.999 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/shutdown => com.twitter.server.handler.ShutdownHandler
    jvm 1    | 15:02:24.999 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/tracing => com.twitter.server.handler.TracingHandler
    jvm 1    | 15:02:24.999 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/events => com.twitter.server.handler.EventsHandler
    jvm 1    | 15:02:24.999 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/events/record/ => com.twitter.server.handler.EventRecordingHandler
    jvm 1    | 15:02:24.999 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/logging => com.twitter.server.handler.LoggingHandler
    jvm 1    | 15:02:24.999 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/metrics => com.twitter.server.handler.MetricQueryHandler
    jvm 1    | 15:02:25.000 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/clients/ => com.twitter.server.handler.ClientRegistryHandler
    jvm 1    | 15:02:25.000 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/servers/ => com.twitter.server.handler.ServerRegistryHandler
    jvm 1    | 15:02:25.000 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/files/ => com.twitter.server.handler.ResourceHandler
    jvm 1    | 15:02:25.000 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/registry.json => com.twitter.server.handler.RegistryHandler
    jvm 1    | 15:02:25.000 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/toggles => com.twitter.server.handler.ToggleHandler
    jvm 1    | 15:02:25.000 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/toggles/ => com.twitter.server.handler.ToggleHandler
    jvm 1    | 15:02:25.000 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /favicon.ico => com.twitter.server.handler.ResourceHandler
    jvm 1    | 15:02:25.001 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/histograms => com.twitter.server.handler.HistogramQueryHandler
    jvm 1    | 15:02:25.001 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - /admin/histograms.json => com.twitter.server.handler.HistogramQueryHandler
    jvm 1    | 15:02:25.020 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - http server started on port: 8888
    jvm 1    | 15:02:25.021 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - Enabling health endpoint on port 9990
    jvm 1    | 15:02:25.022 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - com.deciphernow.server.rest.GMFabricRestServer started.
    jvm 1    | 15:02:25.023 [WrapperSimpleAppMain] INFO  c.d.server.rest.GMFabricRestServer - Startup complete, server ready.
    jvm 1    | My added attribute : 0.0.0
    jvm 1    | 15:02:37.335 [finagle/netty3-1] INFO  c.t.f.h.filters.AccessLoggingFilter - 127.0.0.1 - - [26/Jan/2018:20:02:37 +0000] "GET /ping HTTP/1.1" 200 5 10 "curl/7.55.1"


### Java reference
When referencing the attributes in java casting to the value type is required.

    String version = (String)com.acme.config.version.apply();
    

## Word to the wise
Make sure to instantiate any classes that use attributes, outside of the server, last within the premain for example __MyFirstMicroserviceManager__.

      premain {
    
        //accessManager = new FileWhitelistImpersonationAccessManager(
        //   whitelistFile(), Duration(1, "minute")
        //)
    
        //
        // Decryptor plugin - retrieve instance here.
        // val decryptor = DecryptorManager.getInstance
    
        //
        // Instantiate all business logic after the decryptor instantiation.
        myFirstMicroserviceManager = new MyFirstMicroserviceManager
      }

If classes are instantiated before the __premain__ and their attributes are being called within the code, they will always return the __DEFAULT__ value. 

documentation/Parameters.md
# Overview
The following parameters can be configured in the `etc/parameters.config` file. To create your own attributes and adding them to __parameters.config__ ![see](Config.md).

## White-list Enforcement File
This points to the file that identifies the services that are allowed to access and perform impersonation.

    -acl.whitelist.file=etc/whitelist.txt
    
## Increase the Size of Netty client worker pool
    
    -com.deciphernow.server.config.client.numWorkers=16
    
## Turn on use separate thread pool for clients
The system defaults to true.
    
    -com.deciphernow.server.config.client.useSeparatePool=true
    
## Keystore
If passwords are encrypted, See __Enable a Decryptor__
    
    -com.deciphernow.server.config.tls.keyStore=etc/keystore.jks
    -com.deciphernow.server.config.tls.keyStorePass=password
    
## Truststore
If passwords are encrypted, See __Enable a Decryptor__

    -com.deciphernow.server.config.tls.trustStore=etc/truststore.jks
    -com.deciphernow.server.config.tls.trustStorePass=password
    
## Enable a Decryptor
If not using any encrypted passwords for keystore/truststore then this attribute is not required. If passwords
are encrypted through your implementation then enable as shown.

    -com.deciphernow.server.config.resources.decryptClass=com.blah.my.own.Decryptor
    
## Changing the Maximum Request Size
The default value is **25** Megabytes. It is a long value always represented as **Megabytes**. It must be less than **2**GB. You don't have to specify the unit of measure.

    -com.deciphernow.server.config.rest.maxRequestSize=25

### Bind and Announce
To configure different ports to __BIND__ to or __ANNOUNCE__ different ports [see](AnnounceAndBind.md).
    
## Set the Thrift method level statistics unit of measure

The following units of measure are available: `NANOSECONDS`, `MICROSECONDS`, `MILLISECONDS`, `SECONDS`, `MINUTES`, `HOURS`, `DAYS`.
Unfortunately the unit of measure is not displayed in /admin/metrics.json and thus there can be a mixture of Units Of Measure which could lead to confusion.

    -com.deciphernow.server.config.stats.time=MILLISECONDS

## Announce the Thrift and RESTful endpoints to Zookeeper
    
    -com.deciphernow.server.config.zk.announcementPoint=/example/service/mywombat/1.0
    -com.deciphernow.server.config.zk.zookeeperConnection=localhost:2181
    
## Enable IP address resolution instead of DNS
If `enableIpAddressResolution` is true and no ethernet interface name assigned to `userNetworkInterfaceName` it will iterate over all interface names looking for a valid IPv4 that is NOT a Loopback.

    -com.deciphernow.server.config.ipAddress.enableIpAddressResolution=false
    -com.deciphernow.server.config.ipAddress.useNetworkInterfaceName=eth0
 
# Ignore creating StatsReceiver for static routes.
The framework auto-magically understands how to create the StatsReceiver for the static route. No need to create it.
   
    -com.deciphernow.server.config.staticRoutes.ignore=/ping

## Assigning ENVIRONMENT variables
For all attributes that are of type __String__ an ENVIRONMENT variable may be defined as in __Example 1__.

Example 1

    -com.deciphernow.server.config.zk.announcementPoint=ENV_ZK_ANNOUNCE
    
To retrieve the value assigned to the __zk.announcementPoint__ you would do the following ([based having followed](CreatingNewMS.md)):


    package com.acme
    
    import com.deciphernow.server.ConfigUtilities
    
    /**
     *
     */
    class MyFirstMicroserviceManager {
    
      /**
        * Keep this. Use to see if the service is accepting connections.
        * @return
        */
      def getPong = {
        val ap = ConfigUtilities.selectValue(com.deciphernow.server.config.zk.announcementPoint.apply))
        "pong\n"
      }
    
    }

documentation/ReleaseNotes.md

# Release Notes

## gm-fabric-jvm-0.2.4

### Issue 112

- Using 'enableIpAddressResolution' was not registering IP address if DNS resolution is available.

## gm-fabric-jvm-0.2.3

* Key is that attributes can now be referenced across all modules. No need to include a separate set of properties to be loaded by a Java properties loader.
* The ability to assign environment variable to all String based attributes.
* Improved documentation.
* RestController tests get generated in a microservice now.

### Issue 101

- Framework adds references to `core` && `config` to all modules of a generated service.

### Issue #99

- Parameters that are __String__ defined can be set by a system environment variable.

### Issue #93/94

- Document all changes to framework.

### Issue #92

- Update parameters.config to have all available parameters and documentation.

### Issue #91

- Add Unit Test to RestController in a generated microservice.

### Issue #89

- Fix plugin dependency management.

### Issue #85

- Remove post RPM install script.

### Issue #84

- Move Config out of core into it's own module called 'config'

-------
## gm-fabric-jvm-0.2.2

### Issue #80

- Fixed cloudwatch plugin to acquire memory values though MXBean.

### Issue #79

- Enhanced dependency management.

### Issue #77

- Added documentation for cloudwatch plugin.

### Issue #72

- Updated integration tests to work with CircleCi.

### Issue # 71

- Migration of the cloudwatch plugin into this project.

### Issue #67

- Improved documentation with explicit requirements for local dev environment.

-------

## gm-fabric-jvm-0.2.1

- Some files in 0.2.0 release did not get versioned correctly.

-------

## gm-fabric-jvm-0.2.0

### Issue #55

- Dockerized services announce container IP instead of host IP. 
   Also required capability to override announcment ports.
   
   See [AnnounceAndBind](AnnounceAndBind.md)
   
### Issue #51

- REST service doesn't announce IP to Zookeeper when using 'enableIpAddressResolution' parameters flag.

-------

## gm-fabric-jvm-0.1.4

### Issue #42

- Fix thrift service discovery in zookeeper by removing postfix identifier to 'thrift' registration.

### Issue #11

- Service installs with __Major.Minor__ API version.
- Assigned default install directory to '/opt/services'.
- Assigned default user 'srvcuser'.
- RPM name generated with the __Major.Minor__ as well as handling 'SNAPSHOT' with 'BuildNumber'.


config/src/main/scala/com/deciphernow/server/config/Config.scala
package com.deciphernow.server.config


/*
    Copyright 2017 Decipher Technology Studios LLC

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

import com.twitter.app.GlobalFlag
import com.twitter.app.GlobalFlagVisible

/**
  *
  */
package resources {
  object decryptClass extends GlobalFlag[String]("com.deciphernow.server.support.NoDecryptor","Class to decrypt data")
}


/**
  * Environment variables that override the configuration in parameters.config.
  *
  */
package os.env {
  import com.deciphernow.server.Implicits.stringOptionFlaggable
  object hostname extends GlobalFlag[Option[String]](None, "Environment variable name that will contain the 'hostname' or 'ipAddress'")
  object adminPort extends GlobalFlag[Option[String]](None, "Environment variable name that will contain the admin port.")
  object httpPort extends GlobalFlag[Option[String]](None, "Environment variable name that will contain the http port.")
  object httpsPort extends GlobalFlag[Option[String]](None, "Environment variable name that will contain the https port.")
  object thriftPort extends GlobalFlag[Option[String]](None, "Environment variable name that will contain the thrift port.")
}

/**
  * Define the RESTful ports.
  */
package rest {

  object httpPort  extends GlobalFlag[String](":8888", "HTTP Port")
  object httpsPort extends GlobalFlag[String](":8999", "HTTPS Port")
  object maxRequestSize extends GlobalFlag[Long](25, "Default unit of measure is Megabytes. Must be less than 2 GB!")
}

/**
  * Define the Admin port. It does not follow Twitter in that the real AdminHttp port is an
  * integer.
  *
  */
package admin {

  object port extends GlobalFlag[String](":9990", "Admin Port")

}

/**
  * Define the Thrift port.
  */
package thrift {

  object port extends GlobalFlag[String](":9090", "Thrift Port")
}

/**
  * Configure the stats filter level of time.
  */
package stats {
  object time extends GlobalFlag[String]("MILLISECONDS", "Set time to milliseconds.")
}

/**
  * Define Zookeeper Information.
  */
package zk {

  object announcementPoint extends GlobalFlag[String]("", "Service ZK Announcement Point")
  object zookeeperConnection extends GlobalFlag[String]("", "Zookeeper Connection String")

}

/**
  * Define TLS information for configuring 2-Way SSL.
  */
package tls {

  import com.deciphernow.server.Implicits.fileOptionFlaggable
  import java.io.File

  object keyStore extends GlobalFlag[Option[File]](None, "SSL Key Store File")
  object keyStorePass extends GlobalFlag[String]("", "SSL Key Store Password")
  object trustStore extends GlobalFlag[Option[File]](None, "SSL Trust Store File")
  object trustStorePass extends GlobalFlag[String]("", "SSL Trust Store password")

}

/**
  * Configure the number of workers & separate thread pool.
  */
package client {

  import com.twitter.jvm.numProcs

  object numWorkers extends GlobalFlag[Int]((numProcs() * 2).ceil.toInt, "Size of netty3 client worker pool")
  object useSeparatePool extends GlobalFlag[Boolean](true, "Whether to use a separate thread pool for clients")
}

/**
  * Tell the framework to register IP addresses to Zookeeper instead of it's hostname.
  */
package ipAddress {
  object enableIpAddressResolution extends GlobalFlag[Boolean](false, "The announce IP address instead of hostname. Default is false")
  object useNetworkInterfaceName extends GlobalFlag[String]("", "The network interface to use to find the ip address. Defaults to ''")
}

package staticRoutes {
  object ignore extends GlobalFlag[String]("/ping", "This is a comma separated list of URI that the system will automatically register with a StatsReceiver so don't register these.")
}

package filter {
  object blockHttp extends GlobalFlag[Boolean](false,"The AclRestFilter blocks HTTP requests. Default is to allow HTTP requests.")
}
